bus-mapping/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+//! Crate docs
+
+#![cfg_attr(docsrs, feature(doc_cfg))]
+// Temporary until we have more of the crate implemented.
+#![allow(dead_code)]
+// Catch documentation errors caused by code changes.
+#![deny(broken_intra_doc_links)]
+#![deny(missing_debug_implementations)]
+#![deny(missing_docs)]
+#![deny(unsafe_code)]
+
+use std::{
+    collections::BTreeMap,
+    iter::Iterator,
+    ops::{Index, IndexMut},
+};
+
+use itertools::Itertools;
+use pasta_curves::arithmetic::FieldExt;
+
+// -------- EVM Circuit
+// gc as index of each row corresponds to the state after the opcode is performed
+// 1 gc has 1 entry
+//`StackElem{key 	val 	stack_op(READ/WRITE) 	gc   OPCODE    CallID(later_future)}`
+
+//------ State Circuit
+// Group target (stack or memory)
+// Group by key (location in memory/storage)
+// Sorty by gc
+//`MemoryElem{target 	gc 	val1 	val2 	val3}`
+
+#[derive(Debug, Clone)]
+struct BlockConstants<F: FieldExt> {
+    hash: [u8; 256], // Until we know how to deal with it
+    coinbase: F,
+    timestamp: F,
+    number: F,
+    difficulty: F,
+    gaslimit: F,
+    chain_id: F,
+}
+
+#[derive(Debug, Clone)]
+enum RW {
+    READ,
+    WRITE,
+}
+
+/// Doc
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Target {
+    /// Doc
+    Memory,
+    /// Doc
+    Stack,
+    /// Doc
+    Storage,
+}
+
+/// Doc
+#[derive(Debug, Clone)]
+pub struct Operation<F: FieldExt> {
+    mem_op: RW,
+    target: Target,
+    key: usize,
+    value: F,
+    opcode_info: String,
+}
+
+/// Bus Mapping structure
+#[derive(Debug, Clone)]
+pub struct BusMapping<F: FieldExt> {
+    entries: Vec<Operation<F>>,
+    block_ctants: BlockConstants<F>,
+    // Helper to sort by key groups. We store how many different keys we have when building the Bus Mapping
+    mem_ops_sorted: BTreeMap<usize, Vec<Operation<F>>>,
+    stack_ops_sorted: BTreeMap<usize, Vec<Operation<F>>>,
+    storage_ops_sorted: BTreeMap<usize, Vec<Operation<F>>>,
+}
+
+impl<F: FieldExt> Index<usize> for BusMapping<F> {
+    type Output = Operation<F>;
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.entries[index]
+    }
+}
+
+impl<F: FieldExt> IndexMut<usize> for BusMapping<F> {
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        &mut self.entries[index]
+    }
+}
+
+impl<F: FieldExt> From<(Vec<Operation<F>>, BlockConstants<F>)> for BusMapping<F> {
+    fn from(inp: (Vec<Operation<F>>, BlockConstants<F>)) -> Self {
+        // Initialize the BTreeMaps with empty vecs for each key group
+        let mut mem_ops_sorted = BTreeMap::new();
+        let mut stack_ops_sorted = BTreeMap::new();
+        let mut storage_ops_sorted = BTreeMap::new();
+        inp.0
+            .iter()
+            .map(|op| op.key)
+            .unique()
```

Comment:
We could have 2 identical operation `(mem_op, target, key, value, opcode_info)` with different `global_counter` in state circuit. So I don't think we need `unique` here.

Reply:
I do actually think so.

I need a way to initialize the `BTreeMap`s with the keys we're going to group over our Operations.
To do so, I need to know first which are the keys that are unique, otherways I would repeat them on which case the insertion would fail.

To put it on another way:

If I have the following set of Operations:
```
pc  op              stack (top -> down)                  memory
--  --------------  ----------------------------------   ---------------------------------------  
... 
53  JUMPDEST        [    ,          ,           ,    ]   {40: 80,  80:          ,  a0:         }
54  PUSH1 40        [    ,          ,           ,  40]   {40: 80,  80:          ,  a0:         }
56  MLOAD           [    ,          ,           ,  80]   {40: 80,  80:          ,  a0:         }
57  PUSH4 deadbeaf  [    ,          ,   deadbeef,  80]   {40: 80,  80:          ,  a0:         }
62  DUP2            [    ,        80,   deadbeef,  80]   {40: 80,  80:          ,  a0:         }
63  MSTORE          [    ,          ,           ,  80]   {40: 80,  80:  deadbeef,  a0:         }
64  PUSH4 faceb00c  [    ,          ,   faceb00c,  80]   {40: 80,  80:  deadbeef,  a0:         }
69  DUP2            [    ,        80,   faceb00c,  80]   {40: 80,  80:  deadbeef,  a0:         }
70  MLOAD           [    ,  deadbeef,   faceb00c,  80]   {40: 80,  80:  deadbeef,  a0:         }
71  ADD             [    ,          ,  1d97c6efb,  80]   {40: 80,  80:  deadbeef,  a0:         }
72  DUP2            [    ,        80,  1d97c6efb,  80]   {40: 80,  80:  deadbeef,  a0:         }
73  MSTORE          [    ,          ,           ,  80]   {40: 80,  80: 1d97c6efb,  a0:         }
74  PUSH4 cafeb0ba  [    ,          ,   cafeb0ba,  80]   {40: 80,  80: 1d97c6efb,  a0:         }
79  PUSH1 20        [    ,        20,   cafeb0ba,  80]   {40: 80,  80: 1d97c6efb,  a0:         }
81  DUP3            [  80,        20,   cafeb0ba,  80]   {40: 80,  80: 1d97c6efb,  a0:         }
82  ADD             [    ,        a0,   cafeb0ba,  80]   {40: 80,  80: 1d97c6efb,  a0:         }
83  MSTORE          [    ,          ,           ,  80]   {40: 80,  80: 1d97c6efb,  a0: cafeb0ba}
84  POP             [    ,          ,           ,    ]   {40: 80,  80: 1d97c6efb,  a0: cafeb0ba}
...
```

I would need to get  the following unique keys: 
`[1023, 1022, 1021, 0x40, 0x80, 0xa0....]` and with this I will initialize the trees.
Each tree will contain only the ops is related to and cal already group them since each unique key links to a `Vec<Operation<F>>` which acts as value. 

Finally when we ask for the iterator over them, we might or might not apply a `rev()` and finally we `faltten()` making all of the empty keys disappear.

Am I missing anything here??

Reply:
You are right, sorry that I missed the `flatten` in the end.